Dockerfile: use clang to build dockerd/docker-proxy

crazy-max opened the pull request:
follow-up https://github.com/moby/moby/pull/44812 and https://github.com/moby/moby/pull/43529

**- What I did**

Atm static binaries for dockerd are broken on armhf and armel (32-bit):

```
$ cat /proc/cpuinfo 
processor       : 0
model name      : ARMv7 Processor rev 3 (v7l)
BogoMIPS        : 108.00
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32 
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd08
CPU revision    : 3

processor       : 1
...

Hardware        : BCM2711
Revision        : c03111
Serial          : 10000000d198f4cb
Model           : Raspberry Pi 4 Model B Rev 1.1

$ uname -a
Linux rasp4 5.15.32-v7l+ #1538 SMP Thu Mar 31 19:39:41 BST 2022 armv7l GNU/Linux

$ ./dockerd --version
Illegal instruction
```

```
(gdb) run
Starting program: /home/foo/dockerd 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/arm-linux-gnueabihf/libthread_db.so.1".

Program received signal SIGILL, Illegal instruction.
0x018cd968 in memcpy_ifunc ()
(gdb) bt
#0  0x018cd968 in memcpy_ifunc ()
#1  0x018b89da in __libc_start_main ()
#2  0x000103ec in _start ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb) disassemble
Dump of assembler code for function memcpy_ifunc:
   0x018cd95c <+0>:     lsls    r3, r0, #19
   0x018cd95e <+2>:     bmi.n   0x18cd966 <memcpy_ifunc+10>
   0x018cd960 <+4>:     ldr     r0, [pc, #8]    ; (0x18cd96c <memcpy_ifunc+16>)
   0x018cd962 <+6>:     add     r0, pc
   0x018cd964 <+8>:     bx      lr
   0x018cd966 <+10>:    ldr     r0, [pc, #8]    ; (0x18cd970 <memcpy_ifunc+20>)
=> 0x018cd968 <+12>:    add     r0, pc
   0x018cd96a <+14>:    bx      lr
   0x018cd96c <+16>:    andeq   r0, r0, r10, asr r6
   0x018cd970 <+20>:    andeq   r0, r0, r4, asr r1
End of assembler dump.
```

It seems to be an issue with GCC as building using clang solves this issue:

```
$ ./dockerd --version
Docker version dev, build HEAD
```

```
$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.8.2-docker)

Server:
 Containers: 4
  Running: 4
  Paused: 0
  Stopped: 0
 Images: 29
 Server Version: dev
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 212e8b6fa2f44b9c21b2798135fc6fb7c53efc16
 runc version: v1.1.1-0-g52de29d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.32-v7l+
 Operating System: Raspbian GNU/Linux 11 (bullseye)
 OSType: linux
 Architecture: armv7l
 CPUs: 4
 Total Memory: 3.749GiB
 Name: dns2
 ID: 9fb1c1f4-43ac-4f0e-85ae-c32b5edd899d
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**- How I did it**

~Set `CGO_ENABLED=0` for static build in the Dockerfile. Maybe we should disable CGO only for armel/armhf archs?~

Build dockerd and docker-proxy using clang and also adds extra instruction to prefer ld for cross-compiling arm64 in bullseye otherwise it doesn't link.

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

crazy-max committed:
Dockerfile: smoke tests for static builds

Adds smoke test stage in our Dockerfile and a
GitHub Action workflow

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

thaJeztah commented:
Just a quick check what parts would be excluded if we don't have cgo enabled;

```bash
git grep '//go:build.*cgo'
daemon/graphdriver/btrfs/dummy_unsupported.go://go:build !linux || !cgo
daemon/logger/journald/internal/sdjournal/sdjournal.go://go:build linux && cgo && !static_build && journald
daemon/logger/journald/read.go://go:build linux && cgo && !static_build && journald
daemon/logger/journald/read_test.go://go:build linux && cgo && !static_build && journald
pkg/devicemapper/devmapper.go://go:build linux && cgo
pkg/devicemapper/devmapper_log.go://go:build linux && cgo
pkg/devicemapper/devmapper_wrapper.go://go:build linux && cgo
pkg/devicemapper/devmapper_wrapper_dynamic.go://go:build linux && cgo && !static_build
pkg/devicemapper/devmapper_wrapper_dynamic_deferred_remove.go://go:build linux && cgo && !static_build && !libdm_dlsym_deferred_remove && !libdm_no_deferred_remove
pkg/devicemapper/devmapper_wrapper_dynamic_dlsym_deferred_remove.go://go:build linux && cgo && !static_build && libdm_dlsym_deferred_remove && !libdm_no_deferred_remove
pkg/devicemapper/devmapper_wrapper_no_deferred_remove.go://go:build linux && cgo && !libdm_dlsym_deferred_remove && libdm_no_deferred_remove
pkg/devicemapper/ioctl.go://go:build linux && cgo
pkg/loopback/loopback.go://go:build linux && cgo
quota/projectquota.go://go:build linux && !exclude_disk_quota && cgo
quota/projectquota_unsupported.go://go:build (linux && exclude_disk_quota) || (linux && !cgo) || !linux
quota/testhelpers.go://go:build linux && !exclude_disk_quota && cgo
vendor/github.com/containerd/containerd/plugin/plugin_go18.go://go:build go1.8 && !windows && amd64 && !static_build && !gccgo
vendor/github.com/containerd/containerd/plugin/plugin_other.go://go:build !go1.8 || windows || !amd64 || static_build || gccgo
vendor/github.com/google/certificate-transparency-go/x509/root_cgo_darwin.go://go:build cgo && !arm && !arm64 && !ios
vendor/github.com/google/certificate-transparency-go/x509/root_darwin_armx.go://go:build cgo && darwin && (arm || arm64 || ios)
vendor/github.com/google/certificate-transparency-go/x509/root_nocgo_darwin.go://go:build !cgo
vendor/github.com/moby/sys/mount/mounter_freebsd.go://go:build freebsd && cgo
vendor/github.com/moby/sys/mount/mounter_openbsd.go://go:build openbsd && cgo
vendor/github.com/moby/sys/mount/mounter_unsupported.go://go:build (!linux && !freebsd && !openbsd && !windows && !darwin) || (freebsd && !cgo) || (openbsd && !cgo)
vendor/golang.org/x/sys/cpu/cpu_gccgo_arm64.go://go:build gccgo
vendor/golang.org/x/sys/cpu/cpu_gccgo_s390x.go://go:build gccgo
vendor/golang.org/x/sys/cpu/cpu_gccgo_x86.go://go:build (386 || amd64 || amd64p32) && gccgo
vendor/golang.org/x/sys/cpu/syscall_aix_gccgo.go://go:build aix && gccgo
vendor/golang.org/x/sys/unix/gccgo.go://go:build gccgo && !aix && !hurd
vendor/golang.org/x/sys/unix/gccgo_linux_amd64.go://go:build gccgo && linux && amd64
vendor/golang.org/x/sys/unix/syscall_linux_gccgo_386.go://go:build linux && gccgo && 386
vendor/golang.org/x/sys/unix/syscall_linux_gccgo_arm.go://go:build linux && gccgo && arm
vendor/golang.org/x/sys/unix/zsyscall_aix_ppc64_gccgo.go://go:build aix && ppc64 && gccgo

```

thaJeztah commented:
does this also need `netgo` and `osusergo` options to be set somewhere? (or was this issue because we forgot to set `osusergo`? (I always forget the right combination 😂)

crazy-max commented:
> does this also need `netgo` and `osusergo` options to be set somewhere? (or was this issue because we forgot to set `osusergo`? (I always forget the right combination 😂)

Yes already there for static builds:

```
#21 0.194 ---> Making bundle: binary (in /tmp/bundles/binary)
#21 0.206 Building static /tmp/bundles/binary-daemon/dockerd (linux/arm/v7)...
#21 0.206 ++++ go build -o /tmp/bundles/binary-daemon/dockerd -tags 'netgo osusergo static_build  journald libdm_dlsym_deferred_remove' -installsuffix netgo -ldflags '
#21 0.206               -w      -X "github.com/docker/docker/dockerversion.Version=dev"         -X "github.com/docker/docker/dockerversion.GitCommit=HEAD"      -X "github.com/docker/docker/dockerversion.BuildTime=2023-01-18T11:05:26.000000000+00:00"     -X "github.com/docker/docker/dockerversion.PlatformName="       -X "github.com/docker/docker/dockerversion.ProductName="        -X "github.com/docker/docker/dockerversion.DefaultProductLicense="
#21 0.206
#21 0.206
#21 0.206       -extldflags "-static"
#21 0.206
#21 0.206
#21 0.206       ' github.com/docker/docker/cmd/dockerd
```

thaJeztah commented:
Quick summary of some of the above;

- This change (from the looks of it) means that static builds no longer have `btrfs`, `devicemapper`  and `journald` logging
- It's not 100% clear what exactly causes the issue
- TBD: if this only affects arm32, should we continue compiling static binaries _with_ CGO for 64-bit platforms? (:+1: continues "supporting" these features on those builds, but 👎 would make static binaries inconsistent between 32 and 64 bit platforms)
- ❓ Do we know if `btrfs`, `devicemapper` and  `journald` are fully functional on current static builds? (I know at least for devicemapper we  already had warnings/errors that it should not be used with static builds)
- ❓ do we still need the `static_build` build-tags in device mapper (and other locations) if `static_build` means `!cgo` ?

thaJeztah commented:
/cc @tianon @corhere @neersighted @tonistiigi @cpuguy83 for thoughts on some of the above

corhere commented:
I have a hunch the issue has something to do with [GNU indirect functions](https://sourceware.org/glibc/wiki/GNU_IFUNC) and that the debugger output might be misleading. Also, worth noting that the Raspberry Pi 4 Model B is an [ARM Cortex-A72](https://www.raspberrypi.com/documentation/computers/processors.html#bcm2711) (ARMv8, aarch64) but it appears to be running in 32-bit armv7l mode.

> * ❓ Do we know if `btrfs`, `devicemapper` and  `journald` are fully functional on current static builds? (I know at least for devicemapper we  already had warnings/errors that it should not be used with static builds)

- `btrfs` should be fully functional on cgo-enabled static builds as it only depends on libc and no other userspace library.
- `pkg/devicemapper` links against libdevmapper and has lots of `!static_build` build constraints, so the existing warnings/errors are likely justified.
- The only feature of the `journald` log driver which depends on cgo is read, and it links against libsystemd. That's already constrained to `cgo && !static_build` but the core functionality of logging to journald is implemented in pure Go.

crazy-max commented:
> Also, worth noting that the Raspberry Pi 4 Model B is an [ARM Cortex-A72](https://www.raspberrypi.com/documentation/computers/processors.html#bcm2711) (ARMv8, aarch64) but it appears to be running in 32-bit armv7l mode.

Yes I installed Raspbian 11 32-bit on purpose for testing.

```
$ cat /etc/*release
PRETTY_NAME="Raspbian GNU/Linux 11 (bullseye)"
NAME="Raspbian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"

$ lscpu
Architecture:                    armv7l
Byte Order:                      Little Endian
CPU(s):                          4
On-line CPU(s) list:             0-3
Thread(s) per core:              1
Core(s) per socket:              4
Socket(s):                       1
Vendor ID:                       ARM
Model:                           3
Model name:                      Cortex-A72
Stepping:                        r0p3
CPU max MHz:                     1500,0000
CPU min MHz:                     600,0000
BogoMIPS:                        108.00
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Not affected
Vulnerability Spectre v1:        Mitigation; __user pointer sanitization
Vulnerability Spectre v2:        Vulnerable
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
```

Also got a Raspberry 3 Model B but with Ubuntu 20.04 64-bit:

```
$ cat /etc/*release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.4 LTS"
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

$ lscpu
Architecture:                    aarch64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
CPU(s):                          4
On-line CPU(s) list:             0-3
Thread(s) per core:              1
Core(s) per socket:              4
Socket(s):                       1
Vendor ID:                       ARM
Model:                           4
Model name:                      Cortex-A53
Stepping:                        r0p4
CPU max MHz:                     1200.0000
CPU min MHz:                     600.0000
BogoMIPS:                        38.40
L1d cache:                       128 KiB
L1i cache:                       128 KiB
L2 cache:                        512 KiB
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Not affected
Vulnerability Spectre v1:        Mitigation; __user pointer sanitization
Vulnerability Spectre v2:        Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fp asimd evtstrm crc32 cpuid
```

crazy-max commented:
As discussed with @tonistiigi I switched to build with clang and it seems to work so far.

thaJeztah commented:
Oh! Nice. LOL .. just when I wrote up some thoughts (let me post them anyway 😂)

So my (quick) thoughts;

- arm32 static builds have been broken since 20.10 (and because of that we don't have the official `docker` image for it)
- in general, we consider arm32 (rPI) static builds a "minor" share of users, so it's not _top_ of the list
- but: if we can make it work, I'd be very happy, and having a functional build gives "100% more functionality than before"

for `devicemapper`:

- devicemapper's main audience is CentOS7/RHEL7 (which in the past didn't support overlay).
- we deprecated devicemapper (but kept it for now to allow transitioning and/or users to migrate to containerd's devicemapper implementation in future)
- it was not recommended (or broken) on static builds, so not having it is perfectly acceptable
- ^^ I'd be even good to explicitly compile static binaries with `no_devmapper`

For `btrfs` and `journald`; This is more tricky;

- while I don't think the static binaries should be used for "production" scenarios (better use `rpm` or `deb`), if we _did_ support this in the old builds (for 64-bit), we should try to keep that functionality.
- ^^ from that perspective; if we only need this fix for 32-bit, perhaps we should only disable CGO on 32-bit

crazy-max committed:
Dockerfile: use clang to build dockerd/docker-proxy

Static binaries for dockerd are broken on armhf and armel (32-bit).
It seems to be an issue with GCC as building using clang solves
this issue. Also adds extra instruction to prefer ld for
cross-compiling arm64 in bullseye otherwise it doesn't link.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

thaJeztah reviewed (approved):
LGTM
